server/src/lib/model/db.ts

```diff
@@ -586,4 +621,58 @@ export default class DB {
       );
     }
   }
+
+  async getPoints(client_id: string, challenge: string) {
+    const [[row]] = await this.mysql.query(
+      `
+      SELECT (bonus + clip_points + vote_points) AS user
+      FROM (
+          SELECT bonus, clip_points, COUNT(votes.id) AS vote_points
+          FROM (
+              SELECT challenger.client_id, start_date, end_date, bonus, COUNT(clips.id) AS clip_points
+              FROM (
+                ${getParticipantSubquery('self')}
+              ) challenger
+              LEFT JOIN clips ON challenger.client_id = clips.client_id AND clips.created_at BETWEEN start_date AND end_date
+              GROUP BY challenger.client_id, start_date, end_date, bonus
+          ) speaker
+          LEFT JOIN votes ON speaker.client_id = votes.client_id AND votes.created_at BETWEEN start_date AND end_date
+          GROUP BY speaker.client_id, speaker.bonus, speaker.clip_points
+      ) voter
+      `,
+      [client_id, challenge]
+    );
+    return row;
+  }
+
+  async getWeeklyProgress(client_id: string, challenge: string) {
+    const [[row]] = await this.mysql.query(
+      `
+      SELECT speaker.client_id, start_date, end_date, week, teammate_count, clip_count, COUNT(votes.id) AS vote_count
+      FROM (
+          SELECT user.client_id, start_date, end_date, week, teammate_count, COUNT(clips.id) AS clip_count
+          FROM (
+              SELECT user_clients.client_id,
+                  TIMESTAMPDIFF(WEEK, start_date, NOW()) + 1 AS week,
+                  TIMESTAMPADD(WEEK, TIMESTAMPDIFF(WEEK, start_date, NOW()), start_date) AS start_date,
+                  TIMESTAMPADD(WEEK, TIMESTAMPDIFF(WEEK, start_date, NOW()) + 1, start_date) AS end_date,
```

Comment:
Another important change @phirework just pointed out: we should probably be using `UTC_TIMESTAMP()` instead of `NOW()` in all of these queries, since `NOW()` is based on the server's local timezone and won't always be consistent. We may also want to add our own (consistent) time offset based on where we want the challenge dates to fall.

Thoughts?

Reply:
It is a good point, it’s true that NOW() is based on the DB server's local timezone, but I use NOW() for reasons:
1) Other sqls are using NOW(), it may take some effort to consider them all.
2) I suppose we only have one such server in production. And every time related operations are carried out on that server.
3) I think the UI doesn’t have any place to show the time, currently, unless I didn’t see that part in UI.